components/action-mgt/org.wso2.carbon.identity.action.mgt/src/main/java/org/wso2/carbon/identity/action/mgt/ActionManagementServiceImpl.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.action.mgt;
+
+import org.wso2.carbon.identity.action.mgt.constant.ActionMgtConstants;
+import org.wso2.carbon.identity.action.mgt.dao.impl.ActionManagementDAOImpl;
+import org.wso2.carbon.identity.action.mgt.dao.impl.CacheBackedActionMgtDAO;
+import org.wso2.carbon.identity.action.mgt.exception.ActionMgtClientException;
+import org.wso2.carbon.identity.action.mgt.exception.ActionMgtException;
+import org.wso2.carbon.identity.action.mgt.model.Action;
+import org.wso2.carbon.identity.action.mgt.model.TypeEnums;
+import org.wso2.carbon.identity.action.mgt.util.ActionManagementUtil;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Action management service.
+ */
+public class ActionManagementServiceImpl implements ActionManagementService {
+
+    private static final ActionManagementService INSTANCE = new ActionManagementServiceImpl();
+
+    private static final CacheBackedActionMgtDAO CACHE_BACKED_DAO =
+            new CacheBackedActionMgtDAO(new ActionManagementDAOImpl());
+
+    private ActionManagementServiceImpl() {
+    }
+
+    public static ActionManagementService getInstance() {
+
+        return INSTANCE;
+    }
+
+    @Override
+    public Action addAction(String actionType, Action action, String tenantDomain)
+            throws ActionMgtException {
+
+        String resolvedActionType = getActionTypeFromPath(actionType);
+        // Check whether the maximum allowed actions per type is reached.
+        Map<String, Integer> actionsCountPerType = getActionsCountPerType(tenantDomain);
+        if (actionsCountPerType.get(resolvedActionType) != null &&
+                actionsCountPerType.get(resolvedActionType) >= IdentityUtil.getMaximumActionsPerActionType()) {
+            throw ActionManagementUtil.handleClientException(
+                    ActionMgtConstants.ErrorMessages.ERROR_MAXIMUM_ACTIONS_PER_ACTION_TYPE_REACHED);
+        }
```

Comment:
better move this validation to a method for readability

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/5771/commits/7bfd97e7ae0093852196c21b352d6c0f58bae6ac